### PR TITLE
OSOE-69: SCSS-building NPM script instead of Gulp task

### PR DESCRIPTION
**Pull request by 0liver**
OSOE-69

### CODE REVIEW

File: Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    <Folder Include="wwwroot\css\" />
+    <Folder Include="wwwwNonDefaultCssFolder\" />
+  </ItemGroup>
+
+  <Import Condition="'$(NuGetBuild)' != 'true'" Project="..\..\Lombiq.Npm.Targets\Lombiq.Npm.Targets.props" />
+  <Import Condition="'$(NuGetBuild)' != 'true'" Project="..\..\Lombiq.Npm.Targets\Lombiq.Npm.Targets.targets" />
+  <Import Condition="'$(NuGetBuild)' != 'true'" Project="..\Lombiq.NodeJs.Extensions\Lombiq.NodeJs.Extensions.targets" />
+
+  <ItemGroup Condition="'$(NuGetBuild)' != 'true'">
+    <ProjectReference Include="..\Lombiq.NodeJs.Extensions\Lombiq.NodeJs.Extensions.csproj" />
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj]: The sample won't be published to NuGet, so no need for the `NuGetBuild` checks.

File: Lombiq.NodeJs.Extensions/package.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    "stylelint": "^14.5.1",
+    "stylelint-config-standard-scss": "^3.0.0"
+  },
+  "scripts": {
+    "echo-env": "npm run env",
+    "build:styles": "SCSS_SOURCE=Assets/Styles SCSS_TARGET=wwwroot/css pnpm _build:styles",
+    "build:styles:args": "SCSS_SOURCE=$npm_config_source SCSS_TARGET=$npm_config_target pnpm _build:styles",
+    "_build:styles": "run-p _lint:styles _compile:styles",
+    "lint:styles": "SCSS_SOURCE=Assets/Styles SCSS_TARGET=wwwroot/css pnpm _lint:styles",
+    "lint:styles:args": "SCSS_SOURCE=$npm_config_source SCSS_TARGET=$npm_config_target pnpm _lint:styles",
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions/package.json]: Great addition!

File: Docs/styles.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# Scripts for SCSS files
```
Comment (Piedone) [inline Docs/styles.md]: Capitalize file names.

File: Readme.md [hunk 4]
```diff
@@ -1,21 +1,37 @@
-# Lombiq <add project name here>
+# Lombiq NodeJs Extensions
```
Comment (Piedone) [inline Readme.md]: This seems to be the official name when used in a human-readable context, place change everywhere:
```suggestion
# Lombiq Node.js Extensions
```

File: Readme.md [hunk 5]
```diff
@@ -1,21 +1,37 @@
...
-Add detailed documentation here. If it's a lot of content then create documentation pages under the *Docs* folder and link pages here.
+
+## Getting started
+
+Once you've added this project to your solution, add it as an `npm` package to your own project running:
+
+```
+npm install --save-dev ../path/to/Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions
+```
+
```
Comment (Piedone) [inline Readme.md:52]: Add a section similar to https://github.com/Lombiq/Gulp-Extensions#gulp-tasks and link to the other Readmes.

File: Readme.md [hunk 6]
```diff
@@ -1,21 +1,37 @@
...
+Do you want to quickly try out this project and see it in action? Check it out, together with its accompanying [samples](../Lombiq.NodeJs.Extensions.Samples) project, in our [Open-Source Orchard Core Extensions](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions) full Orchard Core solution and also see our other useful Orchard Core-related open-source projects!
 
-Add detailed documentation here. If it's a lot of content then create documentation pages under the *Docs* folder and link pages here.
+
+## Getting started
+
+Once you've added this project to your solution, add it as an `npm` package to your own project running:
+
+```
+npm install --save-dev ../path/to/Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions
```
Comment (Piedone) [inline Readme.md]: Add a note about where the recommended folder of the project is (under Utilities) and use that in the command.

File: Readme.md [hunk 7]
```diff
@@ -1,21 +1,37 @@
...
+This project allows you to use predefined build scripts for *SCSS* files without having to manage neither the scripts nor the necessary *npm* packages yourself.
 
+Also see our [NPM MSBuild Targets](https://github.com/Lombiq/NPM-Targets) library which can make NPM package management a lot easier.
 
-## Documentation
+Do you want to quickly try out this project and see it in action? Check it out, together with its accompanying [samples](../Lombiq.NodeJs.Extensions.Samples) project, in our [Open-Source Orchard Core Extensions](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions) full Orchard Core solution and also see our other useful Orchard Core-related open-source projects!
 
-Add detailed documentation here. If it's a lot of content then create documentation pages under the *Docs* folder and link pages here.
+
+## Getting started
```
Comment (Piedone) [inline Readme.md]: MSBuild integration notes are missing, see: https://github.com/Lombiq/Gulp-Extensions#integrating-with-msbuild

File: Docs/styles.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        "build:styles": "npm explore nodejs-extensions -- npm run build:styles:args --source=<path/to/scss> --target=<path-to-css>",
+        "clean": "npm run clean:styles",
+        "clean:styles": "npm explore nodejs-extensions -- npm run clean:styles:args --target=<path-to-css>",
+        "watch": "npm run watch:styles",
+        "watch:styles": "npm explore nodejs-extensions -- npm run watch:styles:args --source=<path/to/scss> --target=<path-to-css>",
+      }
+      ```
+3. For even more examples and a mixed setup of default and non-default paths, please check out our dedicated [Samples](../Lombiq.NodeJs.Extensions.Samples) project.
+
+To see and run all of the defined scripts in the Visual Studio Task Runner Explorer console, you need to install the [NPM Task Runner](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.NpmTaskRunner64) extension for Visual Studio. You can then run the given scripts and inspect any errors and linter rule violations in the attached console.
```
Comment (Piedone) [inline Docs/styles.md]: This is better in the root Readme since it'll be interesting for all scripts.

File: Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj [hunk 9]
```diff
@@ -0,0 +1,25 @@
+<Project Sdk="Microsoft.NET.Sdk.Razor">
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:1]: Why is wwwwNonDefaultCssFolder/marker.txt necessary?

File: Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    <Warning
+      Code="SL01"
+      Text="Stylelint error in Gulp process, check the build output for more information."
+      Condition="$([System.Text.RegularExpressions.Regex]::IsMatch('$(PrebuildOutput)', 'severity level &quot;error&quot;'))" />
+    <Warning
+      Code="SL02"
+      Text="Stylelint warning in Gulp process, check the build output for more information."
+      Condition="$([System.Text.RegularExpressions.Regex]::IsMatch('$(PrebuildOutput)', 'severity level &quot;warning&quot;'))" />
+  </Target>
+
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets]: There's a nondescriptive error instead of the warning:
image

File: Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    Name="StylelintWarningTarget"
+    AfterTargets="NpmDotnetPrebuild"
+    BeforeTargets="Compile">
+    <Warning
+      Code="SL01"
+      Text="Stylelint error in Gulp process, check the build output for more information."
+      Condition="$([System.Text.RegularExpressions.Regex]::IsMatch('$(PrebuildOutput)', 'severity level &quot;error&quot;'))" />
+    <Warning
+      Code="SL02"
+      Text="Stylelint warning in Gulp process, check the build output for more information."
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets]: Not Gulp anymore.

File: Lombiq.NodeJs.Extensions.Samples/package.json [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  "scripts": {
+    "dotnet-prebuild": "pnpm build",
+    "dotnet-postclean": "pnpm clean",
+    "build": "npm-run-all --parallel build:styles build:styles:args",
+    "build:styles": "npm explore nodejs-extensions -- pnpm build:styles",
+    "build:styles:args": "npm explore nodejs-extensions -- pnpm build:styles:args --source=NonDefaultScssFolder --target=wwwwNonDefaultCssFolder",
+    "clean": "npm-run-all --parallel clean:styles clean:styles:args",
+    "clean:styles": "npm explore nodejs-extensions -- pnpm clean:styles",
+    "clean:styles:args": "npm explore nodejs-extensions -- pnpm clean:styles:args --target=wwwwNonDefaultCssFolder",
+    "watch": "npm-run-all --parallel watch:styles watch:styles:args",
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions.Samples/package.json:16]: I get this error:

image

`watch:styles` works.

File: Lombiq.NodeJs.Extensions/package.json [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    "postcss": "^8.4.7",
+    "postcss-cli": "^9.1.0",
+    "rimraf": "^3.0.2",
+    "sass": "^1.49.8",
+    "stylelint": "^14.5.1",
+    "stylelint-config-standard-scss": "^3.0.0"
+  },
+  "scripts": {
+    "echo-env": "npm run env",
+    "build:styles": "SCSS_SOURCE=Assets/Styles SCSS_TARGET=wwwroot/css pnpm _build:styles",
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions/package.json]: How this works in GE is that if you have linter errors, the SCSS compilation still happens, the CSS is produced (and you get a build warning in MSBuild). It's important to keep this behavior, because during development you can choose to ignore linter violations during quick development, you just need to fix them until the end (i.e. the same as for .NET analyzer violations).

File: Lombiq.NodeJs.Extensions/Scripts/scss-all.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
+const path = require('path');
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions/Scripts/scss-all.js]: Is this script used?

File: Stylelint/example.stylelintrc [hunk 15]
```diff
@@ -0,0 +1,8 @@
+// Copy this file to the root directory and modify the path below if necessary. Remove this comment!
+
+{
+  "extends": "./src/Utilities/Lombiq.Gulp.Extensions/Stylelint/lombiq-base.stylelintrc.json",
```
Comment (Piedone) [inline Stylelint/example.stylelintrc]: Change the path.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Readme.md (hunk 7 above)
Comment (0liver) [reply Readme.md]: Somehow, that got lost. I had it in there, initially.

File: Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj (hunk 9 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:1]: It's not 😀

File: Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets (hunk 10 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets]: That part has changed much more than I would have expected. The `stylelint` cli generate different output than the node script, and also does not discern between warnings and errors.

I've chosen to generate Warnings or Errors depending on the build `Configuration`.

File: Lombiq.NodeJs.Extensions/Scripts/scss-all.js (hunk 14 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions/Scripts/scss-all.js]: No.

File: Lombiq.NodeJs.Extensions.Samples/package.json (hunk 12 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions.Samples/package.json:16]: Hm, I don't. Please try again with the latest updates.

File: Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj [hunk 16]
```diff
@@ -14,12 +14,16 @@
...
+  <Import Project="..\..\Lombiq.Npm.Targets\Lombiq.Npm.Targets.targets" />
+  <Import Project="..\Lombiq.NodeJs.Extensions\Lombiq.NodeJs.Extensions.targets" />
 
-  <ItemGroup Condition="'$(NuGetBuild)' != 'true'">
+  <ItemGroup>
     <ProjectReference Include="..\Lombiq.NodeJs.Extensions\Lombiq.NodeJs.Extensions.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <NpmDotnetPrebuildWatchedFiles Include="package.json;Assets/**/*.*;NonDefaultScssFolder/**/*.*"/>
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:26]: Why is this necessary? Is it necessary in all projects using Node Extensions?

File: Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.csproj [hunk 17]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <DefaultItemExcludes>$(DefaultItemExcludes);.git*;node_modules\**</DefaultItemExcludes>
+    <DefaultItemExcludes>$(DefaultItemExcludes);.*;node_modules\**</DefaultItemExcludes>
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.csproj:5]: Why exclude all `.*` files?

File: Readme.md [hunk 18]
```diff
@@ -13,13 +13,42 @@ Also see our [NPM MSBuild Targets](https://github.com/Lombiq/NPM-Targets) librar
...
 
-## Getting started
+## Installation and usage
 
-Once you've added this project to your solution, add it as an `npm` package to your own project running:
+It's recommended to put this project into a folder named _Lombiq.NodeJs.Extensions_ under the _src/Utilities_ folder, but you can put it anywhere. Then add it as an `npm` package to your own project running:
 
 ```
-npm install --save-dev ../path/to/Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions
+npm install --save-dev ../Utilities/Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions
```
Comment (Piedone) [inline Readme.md]: This only goes up to the Modules folder.

File: Readme.md [hunk 19]
```diff
@@ -13,13 +13,42 @@ Also see our [NPM MSBuild Targets](https://github.com/Lombiq/NPM-Targets) librar
...
+
+
+### Integrating with MSBuild
+
+If you want to integrate Stylelint into MSBuild builds, then you need to include Lombiq's [NPM MSBuild Targets](https://github.com/Lombiq/NPM-Targets), too. Make sure you have a _package.json_ file with the `dotnet-prebuild` and `dotnet-postclean` scripts as indicated in the repository's readme. In the affected projects, you need to import these files in the `.csproj` file:
+
+```xml
+<Import Project="path\to\Lombiq.Npm.Targets\Lombiq.Npm.Targets.props" />
+<Import Project="path\to\Lombiq.Npm.Targets\Lombiq.Npm.Targets.targets" />
+<Import Project="path\to\Lombiq.Gulp.Extensions\Lombiq.NodeJs.Extensions.targets"/>
```
Comment (Piedone) [inline Readme.md]: There's leftover GE reference here. Also, use the same example path as above.

File: Lombiq.NodeJs.Extensions/package.json [hunk 20]
```diff
@@ -16,22 +16,22 @@
   },
   "scripts": {
     "echo-env": "npm run env",
-    "build:styles": "SCSS_SOURCE=Assets/Styles SCSS_TARGET=wwwroot/css pnpm _build:styles",
-    "build:styles:args": "SCSS_SOURCE=$npm_config_source SCSS_TARGET=$npm_config_target pnpm _build:styles",
-    "_build:styles": "run-p _lint:styles _compile:styles",
+    "build:styles": "SCSS_SOURCE=Assets/Styles SCSS_TARGET=wwwroot/css pnpm _build:styles --silent",
+    "build:styles:args": "SCSS_SOURCE=$npm_config_source SCSS_TARGET=$npm_config_target pnpm _build:styles --silent",
+    "_build:styles": "run-p -c -s _lint:styles _compile:styles",
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions/package.json]: Are there long switch names available, like `--silent`? Use those instead, so it's easier to understand what's happening.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj (hunk 16 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:26]: It's necessary so that changes in the _NonDefaultScssFolder_ will trigger a re-build of this modules assets. It's only necessary in projects that use non-default paths.

Comment (0liver) [reply Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:26]: I'll add it to the documentation.

File: Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.csproj (hunk 17 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.csproj:5]: Because of the addition of _.npmrc_ which triggers a ton of warnings when you open it in VS:

2022-02-28_14h01_09

Also, _.pnpm-debug.log_ appeared in the same folder, so I thought hiding all dot files would be suitable.

Comment (Piedone) [reply Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.csproj:5]: I see, OK then.

File: Readme.md (hunk 19 above)
Comment (0liver) [reply Readme.md]: Sorry for that. 😞

File: Lombiq.NodeJs.Extensions/package.json (hunk 20 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions/package.json]: Yes, there are, They just make the command so much longer. But in principle, I agree with you.

File: Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj (hunk 16 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:26]: Actually, this is very much NPM Targets related. It shouldn't be in NodeJs Extensions' documentation.

Comment (Piedone) [reply Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:26]: So could you please add docs about it then?

Comment (0liver) [reply Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:26]: OK.

File: Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets [hunk 21]
```diff
@@ -1,13 +1,19 @@
...
+    Name="GenerateStylesPipelineErrors"
     AfterTargets="NpmDotnetPrebuild"
     BeforeTargets="Compile">
+    <Exec Command="echo Running GenerateStylesPipelineErrors" />
     <Warning
-      Code="SL01"
+      Code="NX01"
       Text="Stylelint error in Node.js Extensions script, check the build output for more information."
-      Condition="$([System.Text.RegularExpressions.Regex]::IsMatch('$(PrebuildOutput)', 'Command failed with exit code'))" />
+      Condition="$(PrebuildOutput.Contains('Error in styles pipeline: stylelint failed'))" />
```
Comment (Piedone) [inline Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets:12]: This is a lot better than the previous solution!

**Review by Piedone — CHANGES_REQUESTED** [review]
See comment.

File: Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj (hunk 16 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions.Samples/Lombiq.NodeJs.Extensions.Samples.csproj:26]: I've adjusted the documentation and pushed the changes.

File: Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets (hunk 21 above)
Comment (0liver) [reply Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets:12]: Thanks! I am fond of it, too 😀

**Review by Piedone — APPROVED** [review]
